extensions/notebook/src/book/remoteBookController.ts

```diff
@@ -0,0 +1,261 @@
+	private _book: RemoteBook;
+	constructor() {
+	}
+
+	public async setRemoteBook(url: URL, remoteLocation: string, zipURL?: URL, tarURL?: URL): Promise<string[]> {
+		if (remoteLocation === 'GitHub') {
+			this._book = new GitHubRemoteBook(url, zipURL, tarURL);
+		} else {
+			this._book = new SharedRemoteBook(url);
+		}
+		return await this._book.createLocalCopy();
+	}
+
+	public openRemoteBook(book: string) {
+		this._book.openRemoteBook(book);
+	}
+
+	public async getReleases(url: URL): Promise<IReleases[]> {
+		let options = {
+			headers: {
+				'User-Agent': 'request'
+			}
+		};
+		return new Promise<IReleases[]>((resolve, reject) => {
+			request.get(url.href, options, (error, response, body) => {
+				if (error) {
+					return reject(error);
+				}
+
+				if (response.statusCode === 404) {
+					return reject('Resource not found');
+				}
+
+				if (response.statusCode !== 200) {
+					return reject(response.statusCode);
+				}
+				let releases = JSON.parse(body);
+				let bookReleases: IReleases[] = [];
+				if (releases) {
+					let keys = Object.keys(releases);
+					keys = keys.filter(key => {
+						let release = {} as IReleases;
+						try {
+							release.remote_path = url;
+							release.tag_name = releases[key].tag_name;
+							release.tarURL = new URL(releases[key].tarball_url);
+							release.zipURL = new URL(releases[key].zipball_url);
+						}
+						catch (error) {
+							throw (error);
+						}
+						bookReleases.push(release);
+					});
+				}
+				resolve(bookReleases);
+			});
+		});
+	}
+}
+
+export interface IReleases {
+	remote_path: URL;
+	tag_name: string;
+	zipURL: URL;
+	tarURL: URL;
+}
+
+export abstract class RemoteBook {
+	protected _book_name: string;
+	protected _version: string;
+	protected _lang_code: string;
+	protected _local_path: URL;
+	protected _tarURL: URL;
+	protected _zipURL: URL;
+	protected outputChannel: vscode.OutputChannel;
+	protected apiWrapper: ApiWrapper;
+
+	constructor(public remote_path: URL) {
+		this.remote_path = remote_path;
+		this.apiWrapper = new ApiWrapper();
+		this.outputChannel = this.apiWrapper.createOutputChannel(msgTaskName);
+	}
+
+	public async abstract createLocalCopy(): Promise<string[]>;
+
+	public openRemoteBook(selectedBook: string) {
+		vscode.commands.executeCommand('bookTreeView.openBook', selectedBook, false, undefined);
+	}
+
+	public setLocalPath() {
+		// Save directory on User directory
+		if (vscode.workspace.workspaceFolders !== undefined) {
+			// Get workspace root path
+			let folders = vscode.workspace.workspaceFolders;
+			try {
+				this._local_path = new URL(folders[0].uri.fsPath);
+			}
+			catch (error) {
+				throw (error);
+			}
+		} else {
+			//If no workspace folder is opened then path is Users directory
+			try {
+				this._local_path = new URL(utils.getUserHome());
+			}
+			catch (error) {
+				throw (error);
+			}
+		}
+	}
+
+	public loadBooks(dir: string, books: string[]): string[] {
+		fs.readdirSync(dir).forEach(file => {
```

Comment:
Don't use sync functions from fs. Use https://nodejs.org/dist/latest-v10.x/docs/api/fs.html#fs_fspromises_readdir_path_options instead and make this async.

The reason for this is that while a sync function is running nothing else in the javascript engine is running (JS is single threaded). With async functions the execution of this function will stop if you use await - but there are still other tasks that it can run in the background while it's waiting (such as other async functions). So it's better performance wise to give execution back until the fs function completes. 

(do the same for the other fs functions you use in this PR too)